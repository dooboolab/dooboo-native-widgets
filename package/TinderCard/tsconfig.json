{
  "extends": "../../tsconfigPackage.json",
  "compilerOptions": {
    "outDir": "./lib"
  },
  "exclude": ["package", "node_modules"]
}
