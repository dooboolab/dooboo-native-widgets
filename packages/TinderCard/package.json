{
  "name": "@dooboo-ui/tinder-card",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "version": "0.0.1",
  "license": "MIT",
  "postinstall": "tsc",
  "scripts": {
    "build": "rm -rf ./lib && tsc",
    "watch": "tsc -w"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/dooboolab/dooboo-ui.git",
    "directory": "src/components/shared/TinderCard"
  },
  "bugs": {
    "url": "https://github.com/dooboolab/dooboo-ui/issues"
  },
  "homepage": "https://github.com/dooboolab/dooboo-ui#readme",
  "dependencies": {
    "dooboolab-welcome": "^1.1.1"
  },
  "devDependencies": {
    "styled-components": "^4.4.0",
    "typescript": "^3.7.5"
  },
  "peerDependencies": {
    "react": ">=16.8",
    "react-native": ">=0.59",
    "styled-components": ">=4.4.0"
  },
  "publishConfig": {
    "access": "public"
  },
  "keywords": [
    "swtich-toggle",
    "react-hook",
    "react-native",
    "dooboo-ui/native",
    "dooboolab"
  ]
}
