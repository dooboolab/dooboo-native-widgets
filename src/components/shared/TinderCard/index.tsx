import {
  Animated,
  Dimensions,
  LayoutAnimation,
  PanResponder,
  Text,
  UIManager,
  View,
} from 'react-native';
import React, {
  ReactElement,
  RefForwardingComponent,
  forwardRef,
  useCallback,
  useImperativeHandle,
  useMemo,
  useState,
} from 'react';

import styled from 'styled-components/native';

const SCREEN_WIDTH = Dimensions.get('screen').width;
const SWIPE_THRESHOLD = 0.25 * SCREEN_WIDTH;
const SWIPE_OUT_DURATION = 250;

export enum TinderCardDirection {
  RIGHT = 'right',
  LEFT = 'left'
}

export interface TinderCardRef {
  handleCancel: () => void;
  forceSwipe: (direction: TinderCardDirection) => void;
}

interface Props {
  testID?: string;
  onSwipeRight?: (item: any) => void;
  onSwipeLeft?: (item: any) => void;
  onCancel?: () => void;
  data: any[];
  renderCards: (item: any, type?: number) => ReactElement;
  renderNoMoreCards: () => ReactElement;
  renderCardLabel?: (type: number) => ReactElement;
  rotate?: boolean;
  stackSize?: number;
}

const Container = styled.View`
  width: ${SCREEN_WIDTH};
  align-items: center;
`;

const TextArea = styled.View`
  width: ${SCREEN_WIDTH};
  height: 100%;
  z-index: 99;
  position: absolute;
  top: 20;
`;

const NoCard = styled.View`
  width: 100%;
  height: 200px;
  justify-content: center;
  align-items: center;
  padding: 10px;
`;

const TinderCard: RefForwardingComponent<TinderCardRef, Props> = (
  props,
  ref,
): ReactElement => {
  const [cardIndex, setCardIndex] = useState(0);
  const [type, setType] = useState(0);
  const position = useMemo(() => new Animated.ValueXY(), []);

  const resetPosition = (): void => {
    setType(0);
    Animated.spring(position, {
      toValue: { x: 0, y: 0 },
    }).start();
  };

  const onSwipeCompleted = useCallback((direction: TinderCardDirection): void => {
    const { onSwipeLeft, onSwipeRight, data } = props;

    position.setValue({ x: 0, y: 0 });

    UIManager.setLayoutAnimationEnabledExperimental &&
      UIManager.setLayoutAnimationEnabledExperimental(true);
    LayoutAnimation.spring();

    let currentIndex = 0;

    setCardIndex((idx) => {
      currentIndex = idx;
      return idx + 1;
    });

    if (direction === TinderCardDirection.RIGHT) {
      onSwipeRight && onSwipeRight(data[currentIndex]);
    } else {
      onSwipeLeft && onSwipeLeft(data[currentIndex]);
    }
  }, []);

  const forceSwipe = (direction: TinderCardDirection): void => {
    const x = direction === TinderCardDirection.RIGHT ? SCREEN_WIDTH : -SCREEN_WIDTH;
    Animated.timing(position, {
      toValue: { x, y: 0 },
      duration: SWIPE_OUT_DURATION,
    }).start(() => {
      onSwipeCompleted(direction);
      setType(0);
    });
  };

  const _panResponder = useMemo(
    () =>
      PanResponder.create({
        onStartShouldSetPanResponder: () => true,
        onPanResponderMove: (evt, gestureState) => {
          if (gestureState.dx > 0) {
            setType(2);
          } else if (gestureState.dx < 0) {
            setType(1);
          }
          position.setValue({ x: gestureState.dx, y: gestureState.dy });
        },
        onPanResponderRelease: (evt, gesture) => {
          if (gesture.dx > SWIPE_THRESHOLD) {
            forceSwipe(TinderCardDirection.RIGHT);
          } else if (gesture.dx < -SWIPE_THRESHOLD) {
            forceSwipe(TinderCardDirection.LEFT);
          } else {
            resetPosition();
          }
        },
      }),
    [],
  );

  const getCardStyle = (): {} => {
    const rotate = position.x.interpolate({
      inputRange: [-SCREEN_WIDTH * 1.5, 0, SCREEN_WIDTH * 1.5],
      outputRange: ['-45deg', '0deg', '45deg'],
    });

    const rotateStyle = props.rotate ? { transform: [{ rotate }] } : {};

    return {
      ...position.getLayout(),
      ...rotateStyle,
    };
  };

  const _renderCards = (): ReactElement | (ReactElement | null)[] => {
    if (!props.data || cardIndex >= props.data.length) {
      if (props.renderNoMoreCards) {
        return props.renderNoMoreCards();
      } else {
        return <View/>;
      }
    }

    const dataSet = props.data.map((item, idx) => {
      if (idx < cardIndex) return null;
      if (idx === cardIndex) {
        return (
          <Animated.View
            key={`card__${idx}`}
            style={[
              getCardStyle(),
              {
                position: 'absolute',
                top: 0,
                width: SCREEN_WIDTH,
                zIndex: 99,
              },
            ]}
            {..._panResponder.panHandlers}>
            {props.renderCards(item, type)}
            <TextArea>
              {props.renderCardLabel && props.renderCardLabel(type)}
            </TextArea>
          </Animated.View>
        );
      }

      const indexGap = idx - cardIndex;
      const stackSize = props.stackSize || 3;
      const behindHeight =
        indexGap <= stackSize ? 10 * indexGap : 10 * stackSize;

      return (
        <Animated.View
          key={`card__${idx}`}
          style={[
            {
              position: 'absolute',
              width: SCREEN_WIDTH - 10 * indexGap,
              top: behindHeight,
              zIndex: 5,
            },
          ]}>
          {props.renderCards(item)}
        </Animated.View>
      );
    });
    return dataSet.reverse();
  };

  const handleCancel = (): void => {
    if (cardIndex > 0 && props.onCancel) {
      setCardIndex((index) => index - 1);
      props.onCancel();
    }
  };

  useImperativeHandle(ref, () => ({
    forceSwipe,
    handleCancel,
  }));

  return (
    <Container>
      {_renderCards()}
    </Container>
  );
};

const _renderNoMoreCards = (): ReactElement => (
  <NoCard>
    <Text>No more cards</Text>
  </NoCard>
);

TinderCard.defaultProps = {
  onSwipeRight: (): void => {},
  onSwipeLeft: (): void => {},
  onCancel: (): void => {},
  renderNoMoreCards: _renderNoMoreCards,
  rotate: false,
  stackSize: 3,
};

export default forwardRef(TinderCard);
