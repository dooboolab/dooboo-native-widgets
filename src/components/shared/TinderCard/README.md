# TinderCard

[![Npm Version](http://img.shields.io/npm/v/@dooboo-ui/native-edit-text.svg?style=flat-square)](https://npmjs.org/package/@dooboo-ui/native-edit-text)
[![Downloads](http://img.shields.io/npm/dm/@dooboo-ui/native-edit-text.svg?style=flat-square)](https://npmjs.org/package/@dooboo-ui/native-edit-text)

## Preview

![TinderCard](https://user-images.githubusercontent.com/31176502/73652277-e4402000-46c9-11ea-9139-b1e453c95a6e.gif)

## Props

|                        | types                     | description          | required | default |
| ---------------------- | ------------------------- | -------------------- | -------- | ------- |
| testID                 | string                    |                      |          |         |
| errorTestID            | string                    |                      |          |         |
| data                   | T[ ]                      |                      |          |         |
| renderCards            | (item: T) => ReactElement |                      |          |         |
| renderNoMoreCards      | ( ) => ReactElement       |                      |          |         |
| onSwipeRight           | (item: T) => void         |                      |          |         |
| onSwipeLeft            | (item: T) => void         |                      |          |         |
| onCancel               | ( ) => void               |                      |          |         |
| swipeRightLabelElement | ( ) => ReactElement       |                      |          |         |
| swipeLeftLabelElement  | ( ) => ReactElement       |                      |          |         |
| swipeRightLabelStyle   | ViewStyle                 |                      |          |         |
| swipeLeftLabelStyle    | ViewStyle                 |                      |          |         |
| swipeLabelAlignStyle   | ViewStyle                 |                      |          |         |
| containerStyle         | ViewStyle                 |                      |          |         |
| frontCardStyle         | ViewStyle                 |                      |          |         |
| backCardsStyle         | ViewStyle                 |                      |          |         |
| shouldRotate           | ViewStyle                 |                      |          |         |

## Getting started

### Installation

```sh
yarn add @dooboo-ui/native
```

### Import

  ```javascript
  import { TinderCard } from '@dooboo-ui/native';
  import { TinderCardDirection, TinderCardRef } from '@dooboo-ui/native/lib/TinderCard';
  ```

  ##### TinderCard
  - TinderCard Component
  - Example

    ```
    <TinderCard
        testID="tinderCard"
        ref={tinderCard}
        onSwipeRight={handleUnlike}
        onSwipeLeft={handleLike}
        onCancel={handleCancel}
        data={data}
        renderCards={_renderCards}
        renderNoMoreCards={_renderNoMoreCards}
        containerStyle={{ width: 300, height: 500 }}
        shouldRotate
        swipeRightLabelElement={(): ReactElement => <LikeLabel>Like!</LikeLabel>}
        swipeLeftLabelElement={(): ReactElement => 
          <UnlikeLabel>Unlike!</UnlikeLabel>
        }
        swipeLabelAlignStyle={{ justifyContent: 'center', alignItems: 'center' }}
    />
    ```

  ##### TinderCardDirection
  - You can use this by importing on `@dooboo-ui/native/lib/TinderCard` 
  - Structure
      if you put swiping direction on event function, you can use this. For example, `TinderCardDirection.RIGHT` or `TinderCardDirection.LEFT`

      ```
      {
        RIGHT = 'right',
        LEFT = 'left'
      }
      ```

  ##### TinderCardRef
  - This make us to use function in `TinderCard`.

    ```
      {
        handleCancel: () => void;
        forceSwipe: (direction: TinderCardDirection) => void;
      }
    ```
  - Usage
    
    1. Import module `TinderCardRef` 
        First of all, you should import module `TinderCardRef` in `@dooboo-ui/native/lib/TinderCard`.
        ```
        import { TinderCardRef } from '@dooboo-ui/native/lib/TinderCard';
        ```
    2. Allocate `useRef()` having type `TinderCardRef` Object.
        ```
        import React, { useRef } from 'react';

        ...

        const tinderCard = useRef<TinderCardRef>(null);
        ```
    3. Write this where you want to use.
        ```
        <TouchableOpacity
          onPress={(): void => {
            tinderCard.current.forceSwipe(TinderCardDirection.LEFT);
          }}
        >
          Button
        </TouchableOpacity>
        ```

        If you are difficult to understand `useRef`, you can read [this tutorial](https://medium.com/dooboolab/vexposing-child-functions-using-imperativehandle-and-forwardref-69b3687808bb).

#### Declare interface `T` for props `data`.
  ```
  interface Item {
    id: string;
    title: string;
    content: string;
    image: ImageSourcePropType;
  }
  ``` 

### Full code of example

  ```javascript
  import { TinderCard } from '@dooboo-ui/native';
  import { TinderCardDirection, TinderCardRef } from '@dooboo-ui/native/lib/TinderCard';

  interface Item {
    id: string;
    title: string;
    content: string;
    image: ImageSourcePropType;
  }

  function Page(): React.ReactElement {
      const tinderCard = useRef<TinderCardRef>(null);

      return (
        <Container>
          <TinderCard
            testID="tinderCard"
            ref={tinderCard}
            onSwipeRight={handleUnlike}
            onSwipeLeft={handleLike}
            onCancel={handleCancel}
            data={data}
            renderCards={_renderCards}
            renderNoMoreCards={_renderNoMoreCards}
            containerStyle={{ width: 300, height: 500 }}
            shouldRotate
            swipeRightLabelElement={(): ReactElement => <LikeLabel>Like!</LikeLabel>}
            swipeLeftLabelElement={(): ReactElement => <UnlikeLabel>Unlike!</UnlikeLabel>}
            swipeLabelAlignStyle={{ justifyContent: 'center', alignItems: 'center' }}
          />
          <View style={{
            position: 'absolute',
            bottom: 0,
            flexDirection: 'row',
            width: '100%',
            justifyContent: 'space-between',
          }}>

            <ButtonWrapper
              style={{ backgroundColor: '#ff7676' }}
              onPress={(): void => {
                tinderCard.current.forceSwipe(TinderCardDirection.LEFT);
              }}
            >
              <StyledText style={{ fontSize: 15 }}>UNLIKE</StyledText>
            </ButtonWrapper>

            <ButtonWrapper
              onPress={(): void => {
                tinderCard.current.handleCancel();
              }}
            >
              <StyledText style={{ fontSize: 15 }}>UNDO</StyledText>
            </ButtonWrapper>

            <ButtonWrapper
              style={{ backgroundColor: '#44d1a6' }}
              onPress={(): void => {
                tinderCard.current.forceSwipe(TinderCardDirection.RIGHT);
              }}
            >
              <StyledText style={{ fontSize: 15 }}>LIKE</StyledText>
            </ButtonWrapper>
          </View>
        </Container>
      );
  }

  export default Page;
  ```
