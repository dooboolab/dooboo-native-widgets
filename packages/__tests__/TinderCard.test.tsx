import 'react-native';

import * as React from 'react';

import TinderCard from '../TinderCard';
// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

// eslint-disable-next-line
let props: any;
let component: React.ReactElement;
// let testingLib: RenderResult;

const createTestProps = (obj: Record<string, unknown>): Record<string, unknown> => ({
  navigation: {
    navigate: jest.fn(),
  },
  ...obj,
});

describe('[TinderCard] render', () => {
  beforeEach(() => {
    props = createTestProps({});
    component = <TinderCard {...props} />;
  });

  it('renders without crashing', () => {
    const rendered: renderer.ReactTestRendererJSON | null = renderer
      .create(component)
      .toJSON();
    expect(rendered).toMatchSnapshot();
    expect(rendered).toBeTruthy();
  });

  // describe('interactions', () => {
  //   beforeEach(() => {
  //     testingLib = render(component);
  //   });

  //   it('should simulate onClick', () => {
  //     const btn = testingLib.queryByTestId('btn');
  //     act(() => {
  //       fireEvent.press(btn);
  //     });
  //     expect(cnt).toBe(3);
  //   });
  // });
});
