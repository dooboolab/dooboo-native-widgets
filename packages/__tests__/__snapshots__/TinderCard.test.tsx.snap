// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[TinderCard] render renders without crashing 1`] = `
<View
  style={
    Array [
      Object {
        "height": "100%",
        "width": "100%",
      },
      undefined,
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderBottomColor": "black",
          "borderBottomWidth": 1,
          "borderLeftColor": "black",
          "borderLeftWidth": 1,
          "borderRightColor": "black",
          "borderRightWidth": 1,
          "borderTopColor": "black",
          "borderTopWidth": 1,
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "justifyContent": "center",
          "paddingBottom": 10,
          "paddingLeft": 10,
          "paddingRight": 10,
          "paddingTop": 10,
        },
        undefined,
      ]
    }
  >
    <Text>
      No more cards
    </Text>
  </View>
</View>
`;
